org.eclipse.jdt.core.tests.model/src/org/eclipse/jdt/core/tests/model/CircularDependencyModelTests.java

```diff
@@ -0,0 +1,99 @@
+ * Copyright (c) 2025 IBM Corporation and others.
+ *
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.jdt.core.tests.model;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import junit.framework.Test;
+import org.eclipse.core.runtime.NullProgressMonitor;
+import org.eclipse.jdt.core.ICompilationUnit;
+import org.eclipse.jdt.core.IPackageFragment;
+import org.eclipse.jdt.core.IPackageFragmentRoot;
+import org.eclipse.jdt.core.IType;
+import org.eclipse.jdt.core.ITypeHierarchy;
+import org.eclipse.jdt.core.JavaModelException;
+
+/**
+ * Model tests related to circular dependencies.
+ */
+public class CircularDependencyModelTests extends AbstractJavaModelTests {
+
+	private static final String PROJECT_NAME = "CircularDependency";
+
+	private IPackageFragmentRoot sourceFolder;
+
+	public CircularDependencyModelTests(String name) {
+		super(name);
+	}
+	public CircularDependencyModelTests(String name, int apiLevel) {
+		super(name, apiLevel);
+	}
+
+	public static Test suite() {
+		return buildModelTestSuite(CircularDependencyModelTests.class);
+	}
+
+	@Override
+	protected void setUp() throws Exception {
+		super.setUp();
+		createJavaProject(PROJECT_NAME, new String[] { "src" },
+				new String[] { "JCL21_LIB" }, "bin", "25");
+		this.sourceFolder = getPackageFragmentRoot(PROJECT_NAME, "src");
+	}
+
+	@Override
+	protected void tearDown() throws Exception {
+		deleteProject(PROJECT_NAME);
+		super.tearDown();
+	}
+
+	public void testHierarchyOfCircular() throws JavaModelException {
+		IPackageFragment pack1= this.sourceFolder.createPackageFragment("test1", false, null);
+		ICompilationUnit cuC= pack1.createCompilationUnit("C.java",
+		"""
+		package test1;
+		public class C extends B {}
+		""", false, null);
+		pack1.createCompilationUnit("B.java",
+				"""
+				package test1;
+				public class B extends A {}
+				""", false, null);
+		pack1.createCompilationUnit("A.java",
+				"""
+				package test1;
+				public class A extends C {}
+				""", false, null);
+
+		IType type = cuC.getType("C");
+
+		CompletableFuture<ITypeHierarchy> hierarchyFuture = CompletableFuture.supplyAsync(() -> {
+			try {
+				return type.newSupertypeHierarchy(new NullProgressMonitor());
+			} catch (JavaModelException e) {
+				throw new RuntimeException(e);
+			}
+		});
+
+		try {
+			// throws if the operation takes more than 5 seconds
+			hierarchyFuture.get(5, TimeUnit.SECONDS);
```

Comment:
No need to wait for so long.

```suggestion
			hierarchyFuture.get(200, TimeUnit.MILLISECONDS);
```